compiler/next/lib/resolution/resolution-types.cpp

```diff
@@ -17,11 +17,235 @@
+  setupForSymbol(func);
+}
+
+bool ResolutionResultByPostorderID::update(ResolutionResultByPostorderID& keep,
+                                           ResolutionResultByPostorderID& addin)
+{
+  return defaultUpdateVec(keep.vec, addin.vec);
+}
+
+bool FormalActualMap::computeAlignment(const UntypedFnSignature* untyped,
+                                       const TypedFnSignature* typed,
+                                       const CallInfo& call) {
+
+  // untyped must be provided but typed can be nullptr.
+  assert(untyped);
+
+  // create the mapping handling named and default arguments
+
+  // clear the current mapping
+  byFormalIdx.empty();
+  actualIdxToFormalIdx.empty();
+  mappingIsValid = false;
+  failingActualIdx = -1;
+
+  // allocate space in the arrays
+  byFormalIdx.resize(untyped->formals.size());
+  actualIdxToFormalIdx.resize(call.actuals.size());
+
+  // initialize the FormalActual parts from the Formals
+  size_t formalIdx = 0;
+  for (const Formal* formal : untyped->formals) {
+    FormalActual& entry = byFormalIdx[formalIdx];
+    entry.formal = formal;
+    if (typed) {
+      entry.formalType = typed->formalTypes[formalIdx];
+    }
+    entry.hasActual = false;
+    entry.actualIdx = -1;
+
+    formalIdx++;
+  }
+
+  // Match named actuals against formal names in the function signature.
+  // Record successful matches in actualIdxToFormalIdx.
+
+  size_t actualIdx = 0;
+  for (const CallInfoActual& actual : call.actuals) {
+    if (!actual.byName.isEmpty()) {
+      bool match = false;
+      int formalIdx = 0;
+      for (const Formal* formal : untyped->formals) {
+        if (actual.byName == formal->name()) {
+          match = true;
+          FormalActual& entry = byFormalIdx[formalIdx];
+          entry.hasActual = true;
+          entry.actualIdx = actualIdx;
+          entry.actualType = actual.type;
+          actualIdxToFormalIdx[actualIdx] = formalIdx;
+          break;
+        }
+        formalIdx++;
+      }
+
+      // Fail if no matching formal is found.
+      if (match == false) {
+        mappingIsValid = false;
+        failingActualIdx = actualIdx;
+        // TODO: track failure type for error messages
+        return false;
+      }
+    }
+    actualIdx++;
+  }
+
+  // Fill in unmatched formals in sequence with the remaining actuals.
+  // Record successful substitutions
+  formalIdx = 0;
+  actualIdx = 0;
+  for (const CallInfoActual& actual : call.actuals) {
+    if (actual.byName.isEmpty()) {
+      // Skip any formals already matched to named arguments
+      while (byFormalIdx[formalIdx].actualIdx >= 0) {
+        if (formalIdx + 1 >= byFormalIdx.size()) {
+          // too many actuals
+          mappingIsValid = false;
+          failingActualIdx = actualIdx;
+          return false;
+        }
+        formalIdx++;
+      }
+      assert(formalIdx < byFormalIdx.size());
+
+      // TODO: special handling for operators
```

Comment:
I thought the way you set up operator calls would avoid the actual/formal number mismatch?

Reply:
That would be nice :) But it's all future work. I left this comment because the code I was porting over had such special handling.

Reply:
That's what I figured, yeah :)